activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -191,60 +195,76 @@ def restore!
       end
 
       def rollback_records
-        return unless records
-
-        ite = unique_records
+        if records
+          begin
+            ite = unique_records
 
-        instances_to_run_callbacks_on = prepare_instances_to_run_callbacks_on(ite)
+            instances_to_run_callbacks_on = prepare_instances_to_run_callbacks_on(ite)
 
-        run_action_on_records(ite, instances_to_run_callbacks_on) do |record, should_run_callbacks|
-          record.rolledback!(force_restore_state: full_rollback?, should_run_callbacks: should_run_callbacks)
-        end
-      ensure
-        ite&.each do |i|
-          i.rolledback!(force_restore_state: full_rollback?, should_run_callbacks: false)
+            run_action_on_records(ite, instances_to_run_callbacks_on) do |record, should_run_callbacks|
+              record.rolledback!(force_restore_state: full_rollback?, should_run_callbacks: should_run_callbacks)
```

Comment:
We're not exposing `full_rollback?` to our new callbacks, but I'm not sure that's a generally useful piece of information... without digging in to history, this feels like it might be facilitating a historical-compatibility quirk rather than a first-principles desired function.